### DISCUSSION TITLE
14 febbraio 2023 A - esercizio 3

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/basi-di-dati-1#31] [Esami]
E' dato un file di 3.175.250 record. Ogni record occupa 350 byte, di cui 125 per la chiave.

Un blocco contiene 2048 byte. Un puntatore a blocco occupa 4 byte. Si utilizza una organizzazione B-TREE.

3a) Calcolare l'occupazione in blocchi del file principale quando l'albero ha altezza massima.

3b) Calcolare l'occupazione in blocchi del file indice (tutti i livelli) quando l'albero ha altezza massima.

3c) Calcolare il costo di una ricerca quando l'albero ha altezza massima.

**Comment by IonutCicio** [comment]
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/40034429/8d45efa6-bb32-485a-8478-ad58c575442c)

**Reply by CasuFrost** [reply]
L'esercizio richiedendo che l'altezza sia massima, impone che i blocchi del file indice siano riempiti al minimo, nel punto **(b)** fai bene a dividere 2048 per 2 quando calcoli i $P_pBI$, però nel punto **(a)** non dovresti dividere per due, in quanto non è imposto che i blocchi del file principale siano riempiti a metà, bensì possono essere anche riempiti totalmente.\
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/97389762/efef4bd3-c8f4-4663-adb9-a1ff9e7afdd7)
Come vedi nell'immagine,  il riempimento dei blocchi del file principale non influisce sull'altezza dell'albero.

**Reply by IonutCicio** [reply]
Il problema è che se i blocchi del file principale sono riempiti totalmente, l'altezza dell'albero diminuisce (in genere). Il primo disegno che hai fatto è sbagliato perché hai una foglia che non punta a nulla. Levando quella foglia, 

![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/40034429/bc6261d6-3c0b-42f6-a3a1-445c4a07ec07)

Il primo disegno dovrebbe avere 3 livelli, non 4. Facendo i conti

| livello | # blocchi |
|:--:|--|
| 0 | $4$ |
| 1 | $\frac{4}{2} = 2$ |
| 2 | $\frac{2}{2} = 1$ |

Il fatto che ti venga lo stesso numero di livelli in questo esercizio è una conincidenza probabilmente, ma non puoi dire che il numero di livelli non dipende dal numero di blocchi nel file principale. Tu costruisci l'indice proprio in base ai blocchi del file principale _(bottom up)_

**Reply by IonutCicio** [reply]
Pdf 21, slide 56, regola 4. Tutte le foglie devono stare allo stesso livello. Nel primo disegno non va bene perché c'è una foglia che sta a un livello superiore rispetto alle altre.

![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/40034429/48ecdc30-d854-4254-965a-85b821c51359)

**Reply by atyion** [reply]
Scusa il disturbo. Quando fai 8 * 129 + 4 credo ti sia sbagliato. Dovresti fare 8 * 125 + 4, pertanto dovrebbe uscire come risultato complessivo 10.

**Reply by IonutCicio** [reply]
8 * 129 + 4 è giusto, perché nel blocco ho **| P C P C P C P C P C P C P C P C P |** dove P sta per puntatore e C per chiave.
La coppia puntatore + chiave ha dimensione 125 + 4 = 129, e ho 8 coppie **P C**, poi ci devo aggiungere un puntatore alla fine, quindi + 4

**Reply by CasuFrost** [reply]
Però il professore fece questo specifico esercizio in classe il 20 dicembre, e le foglie le contò riempite al 100%, trovando un totale di 79386 blocchi

**Reply by IonutCicio** [reply]
Hai ragione sul fatto che in aula il prof. lo fece così _(ho ricontrollato gli appunti)_, ma il ragionamento per cui le foglie possano essere piene e l'albero avere altezza massima non torna perché ci si ritrova con un B-Tree non valido. Di fatto, il controesempio è proprio quello che hai fatto te. In un albero con altezza massima, i blocchi del file principale devono essere pieni al minimo.

**Reply by CasuFrost** [reply]
È ironico come io stesso per avvalorare la mia tesi abbia mostrato un esempio che la confuta.

**Reply by atyion** [reply]
Hai ragione 😶‍🌫️

**Comment by CasuFrost** [comment]
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/97389762/8c211804-73c4-4623-8ebd-7def55c6c9ac)

**Comment by atyion** [comment]
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/101365532/fe51c11f-dd64-4318-8eab-89a9dce3a855)

**Comment by Warcophyr** [comment]
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/133346488/b06c8901-cade-4117-b7f9-9145f21eb4f8)
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/133346488/d62b20d2-1276-41ea-b28b-642f893986a7)
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/133346488/0e9150a3-b803-4cd2-a802-b527cf8184b4)
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/133346488/195747e8-ddb9-43af-90d0-971390ffe19a)
